### PR TITLE
multi-step-ahead forecast functionality

### PR DESCRIPTION
**Pull request by RLstat**
This pull request added the multi-step-ahead forecast functionality with `model.forecast` function. `model.forecast(horizon=n)` will make forecast for `n` timesteps ahead, by recursively make one-step-ahead forecast and put the predicted value into the input vector to make the next forecast. Illustration figure is in #101

### CODE REVIEW

**Comment by RLstat** [comment]
@ethem-kinginthenorth , I will add an example specifically for this new functionality in a day or two. For how multi-step-forecasting works, I have a illustrative figure in issue #101 .

File: dlpy/model.py [hunk 1]
```diff
@@ -896,7 +903,262 @@ def predict(self, data, text_parms=None, layer_out=None, layers=None, gpu=None,
...
+            None means a new CASTable with random name will be generated.
+            Default: None
+            
+        Returns
+        -------
+        :class:`CASTable`
+
+        """
+        
+        if not isinstance(self.train_tbl, TimeseriesTable):
```
Comment (yg2288) [inline dlpy/model.py]: `self.train_tbl` and `self.valid_tbl` gets set only after calling `Model.fit`, and `self.valid_tbl` may be `None`, so may want to check this

**Comment by ethem-kinginthenorth** [comment]
@RLstat could you please address the comments here. Thanks

File: dlpy/model.py (hunk 1 above)
Comment (RLstat) [reply dlpy/model.py]: @yg2288 , Thanks for spotting it! In the modified merge request, I set `self.train_tbl` and `self.valid_tbl` default to be `None`. if `self.train_tbl` is still `None` when use `model.forecast` (maybe the model has not been fitted and is restored or imported from other sources), the user can manually pass a reference to the training `TimeseriesTable` through the `train_table` argument in the `model.forecast`. These information stored in the training `TimeseriesTable`  is essential for multi-step-ahead timeseries forecast, which is the main purpose of `model.forecast`. Without those information(like time interval, timeid, subsequence length etc. ), you cannot do multi-step-forecast, and should just use `model.predict` for non-timeseries prediction. I added these detailed explanations in error messages and documentations as well. 

`self.valid_tbl` can be `None`, and it won't affect down stream analysis. All the places that uses `self.valid_tbl` can handle the case where it is `None`.

Comment (yg2288) [reply dlpy/model.py]: Awesome, thanks!

**Pull request merged**